### PR TITLE
Add typing for tmt/steps/report/__init__.py

### PR DESCRIPTION
**Pull request by janhavlin**

### CODE REVIEW

File: tmt/steps/report/__init__.py
Comment (psss) [inline tmt/steps/report/__init__.py:97]: What about making the `usage` parameter mandatory instead? Seems more straightforward and it always needs to be set anyway.

File: tmt/steps/report/__init__.py
Comment (psss) [inline tmt/steps/report/__init__.py:106]: @happz, is there a type for `click`'s context object? Or are we good with `Any` here?

Comment (happz) [reply tmt/steps/report/__init__.py:106]: Yep, `click.Context` :)

Every single `Any` needs to be examined, it's like an admission of defeat, and we shall not give up so easily :)

File: tmt/steps/report/__init__.py
Comment (janhavlin) [reply tmt/steps/report/__init__.py:22]: Because method `ReportPlugin.delegate()` expects `Dict[Any, Any]` which is not compatible with `tmt.steps.StepData` (derived from `TypedDict`):

```
tmt/steps/report/__init__.py:22: error: Argument 2 to "delegate" of "Plugin" has incompatible type "StepData"; expected "Dict[Any, Any]"
```

See e.g. https://github.com/python/mypy/issues/4976.

I agree that converting to `dict` is not so pretty. Perhaps changing the method to `def delegate(..., data: Union[Dict[Any, Any], StepData])` could work.

File: tmt/steps/report/__init__.py
Comment (janhavlin) [reply tmt/steps/report/__init__.py:97]: Is this true for all `base_command()` methods? To me it looks like it should be done in its own patch separately from typing annotations since it is changing how the method is used.

File: tmt/steps/report/__init__.py
Comment (happz) [reply tmt/steps/report/__init__.py:91]: What I'm sure about is that the use of `click.Option` in https://github.com/teemtee/tmt/pull/1327/ is likely wrong - that's what we realized last week or so on IRC, while looking into https://github.com/teemtee/tmt/pull/1304. What you're looking for there is `ClickOptionDecoratorType`, that's what all these `options()` methods return.

And since I now know which patch were we chatting about, I can point you to the right type, which should be `Type[click.Command]`, not `click.Command`. IIUIC, `method_class` is supposed to be exactly that, a class, not an instance, which means that you need one step back and "dereference" from "instance of click.Command class" to "type of instance of click.Command", which is done with `Type[...]` annotation.

Note that @thrix is adding this very annotation to `method_class` - mypy raised some complaints and this was needed to amend what `options.py` thins is right and what the rest of tmt code expected.

Comment (janhavlin) [reply tmt/steps/report/__init__.py:91]: > What I'm sure about is that the use of click.Option in https://github.com/teemtee/tmt/pull/1327 is likely wrong.

Ok, thanks for the clarification. That means that all the current `def options(...) -> List[click.Option]:` should be changed. I see it is annotated correctly in #1304 and the other occurrence I found is in [`provision/artemis.py`](https://github.com/teemtee/tmt/blob/main/tmt/steps/provision/artemis.py#L309).

> I can point you to the right type, which should be Type[click.Command], not click.Command.

Yes, sorry, I wrote `click.Command` but I meant `Type[click.Command]` - I understand the difference.

Let's block this PR until #1304 is merged because this patch is depending on some of its changes.

File: tmt/steps/report/__init__.py
Comment (janhavlin) [reply tmt/steps/report/__init__.py]: Updated, thanks!

File: tmt/steps/report/__init__.py
Comment (psss) [reply tmt/steps/report/__init__.py]: Yes, `usage` has to be provided for all commands. Probably not worth of a separate pull request, covered by 809aed5.

**Review by psss — APPROVED** [review]
After changes in ffaa8ec and 809aed5 looks good and ready for merge. @janhavlin, @thrix, could you have a quick look and confirm those changes are ok?

**Pull request merged**
